hoodie-hadoop-mr/src/main/java/com/uber/hoodie/hadoop/realtime/HoodieRealtimeRecordReader.java

```diff
@@ -106,29 +88,30 @@ private void readAndCompactLog() throws IOException {
         // TODO(vc): In the future, the reader schema should be updated based on log files & be able to null out fields not present before
         Schema readerSchema = generateProjectionSchema(writerSchema, projectionFields);
 
-        LOG.info(String.format("About to read logs %s for base split %s, projecting cols %s",
+        LOG.info(String.format("About to read and compact logs %s for base split %s, projecting cols %s",
                 split.getDeltaFilePaths(), split.getPath(), projectionFields));
-        for (String logFilePath: split.getDeltaFilePaths()) {
-            GenericDatumReader<GenericRecord> datumReader = new GenericDatumReader<>(writerSchema, readerSchema);
-            final AvroFSInput input = new AvroFSInput(FileContext.getFileContext(jobConf), new Path(logFilePath));
-            DataFileReader<GenericRecord> reader = (DataFileReader<GenericRecord>) DataFileReader.openReader(input, datumReader);
-            while (reader.hasNext()) {
-                GenericRecord rec = reader.next();
-                String key = rec.get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString();
-                String commitTime = rec.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD).toString();
-                if (HoodieTimeline.compareTimestamps(commitTime, split.getMaxCommitTime(), HoodieTimeline.GREATER)) {
-                    // stop reading this log file. we hit a record later than max known commit time.
-                    break;
-                }
 
-                // we assume, a later safe record in the log, is newer than what we have in the map & replace it.
-                ArrayWritable aWritable = (ArrayWritable) avroToArrayWritable(rec, writerSchema);
-                deltaRecordMap.put(key, aWritable);
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Log record : " + arrayWritableToString(aWritable));
-                }
+        HoodieCompactedLogRecordScanner compactedLogRecordScanner =
```

Comment:
CompactedLogRecordScanner already merges the records and provides the compacted scanner of GenericRecords. You just need to iterate over it and convert it to ArrayWritable. You dont have to merge.